File: android-example/app/src/main/java/com/dsrv/wallet/example/wallet/component/HistorySection.kt
package com.dsrv.wallet.example.wallet.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.AssistChip
import androidx.compose.material3.Card
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.dsrv.wallet.example.wallet.model.TransactionHistoryItem
import com.dsrv.wallet.example.wallet.model.Wallet

/**
 * 거래 내역 조회 — customer-backend `GET /sdk/transactions` (선택 지갑 fromAddress 기준).
 */
@Composable
fun HistorySection(modifier: Modifier = Modifier) {
    val wallet: Wallet = viewModel()
    val state = wallet.uiState
    val address = wallet.addressText

    // 진입 시 자동 조회 (지갑 변경 시 재조회)
    LaunchedEffect(address) {
        if (address.isNotEmpty()) wallet.getTransactionHistory()
    }

    SectionContainer(
        title = "거래 내역",
        subtitle = "총 ${state.historyTotal}건 · 지갑 ${shortHex(address)}",
        modifier = modifier,
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.End,
        ) {
            if (state.historyLoading) {
                CircularProgressIndicator(Modifier.size(14.dp), strokeWidth = 2.dp)
            }
            TextButton(
                onClick = { wallet.getTransactionHistory() },
                enabled = state.sdkInitialized && !state.historyLoading,
            ) { Text("새로고침") }
        }

        state.historyError?.let {
            Text("⚠ $it", color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall)
            Spacer(Modifier.height(6.dp))
        }

        if (state.historyItems.isEmpty() && !state.historyLoading && state.historyError == null) {
            Text(
                "거래 내역이 없습니다.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }

        state.historyItems.forEachIndexed { i, item ->
            HistoryItemCard(item)
            if (i < state.historyItems.lastIndex) Spacer(Modifier.height(8.dp))
        }

        if (state.historyItems.size < state.historyTotal) {
            Spacer(Modifier.height(10.dp))
            OutlinedButton(
                onClick = { wallet.getTransactionHistory(loadMore = true) },
                enabled = !state.historyLoading,
                modifier = Modifier.fillMaxWidth(),
            ) { Text("더 보기 (${state.historyItems.size}/${state.historyTotal})") }
        }
    }
}

@Composable
private fun HistoryItemCard(item: TransactionHistoryItem) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.padding(12.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(item.method ?: "transaction", style = MaterialTheme.typography.titleSmall)
                    Text(
                        item.createdAt,
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                AssistChip(onClick = {}, label = { Text(item.status) })
            }
            Spacer(Modifier.height(6.dp))
            HorizontalDivider()
            Spacer(Modifier.height(6.dp))
            HistoryRow("체인", "${item.chainId} (${item.chainType})")
            HistoryRow("보낸 주소", shortHex(item.fromAddress))
            item.toAddress?.let { HistoryRow("받는 주소", shortHex(it)) }
            HistoryRow("txId", item.transactionId)
            item.txHash?.let {
                Spacer(Modifier.height(4.dp))
                Text("txHash", style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                CopyableText(text = it, singleLine = true)
            }
        }
    }
}

@Composable
private fun HistoryRow(label: String, value: String) {
    Row(modifier = Modifier.padding(vertical = 2.dp)) {
        Text(
            label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.width(76.dp),
        )
        Text(value, style = MaterialTheme.typography.labelMedium)
    }
}

private fun shortHex(s: String): String = if (s.length <= 16) s else "${s.take(10)}…${s.takeLast(4)}"

File: flutter-example/dsrv-wallet-sdk-flutter-example/lib/components/history_section.dart
import 'package:flutter/material.dart';

import '../transaction_history_repository.dart';
import '../ui.dart';
import '../wallet_state.dart';

/// 거래 내역 조회 — customer-backend `GET /sdk/transactions` (선택 지갑 fromAddress 기준).
/// Android `HistorySection.kt` / iOS `HistorySection.swift` 대응.
class HistorySection extends StatefulWidget {
  final WalletState wallet;
  const HistorySection({super.key, required this.wallet});

  @override
  State<HistorySection> createState() => _HistorySectionState();
}

class _HistorySectionState extends State<HistorySection> {
  @override
  void initState() {
    super.initState();
    // 진입 시 자동 조회
    if (widget.wallet.address.isNotEmpty) {
      widget.wallet.getTransactionHistory();
    }
  }

  @override
  Widget build(BuildContext context) {
    final wallet = widget.wallet;
    final loading = wallet.busy('history');

    return SectionCard(
      '거래 내역',
      subtitle: '총 ${wallet.historyTotal}건 · 지갑 ${_shortHex(wallet.address)}',
      children: [
        Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            if (loading)
              const SizedBox(
                width: 14,
                height: 14,
                child: CircularProgressIndicator(strokeWidth: 2),
              ),
            TextButton(
              onPressed: (!wallet.initialized || loading)
                  ? null
                  : () => wallet.getTransactionHistory(),
              child: const Text('새로고침'),
            ),
          ],
        ),
        if (wallet.historyError != null) ErrorLine(wallet.historyError!),
        if (wallet.historyItems.isEmpty &&
            !loading &&
            wallet.historyError == null)
          Text('거래 내역이 없습니다.',
              style: TextStyle(fontSize: 12, color: Theme.of(context).hintColor)),
        for (final item in wallet.historyItems) _HistoryItemCard(item: item),
        if (wallet.historyItems.length < wallet.historyTotal)
          OutlinedButton(
            onPressed: loading
                ? null
                : () => wallet.getTransactionHistory(loadMore: true),
            child: Text(
                '더 보기 (${wallet.historyItems.length}/${wallet.historyTotal})'),
          ),
      ],
    );
  }
}

class _HistoryItemCard extends StatelessWidget {
  final TransactionHistoryItem item;
  const _HistoryItemCard({required this.item});

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(12),
      decoration: BoxDecoration(
        color: Theme.of(context).colorScheme.surfaceContainerHighest,
        borderRadius: BorderRadius.circular(10),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(item.method ?? 'transaction',
                        style: const TextStyle(
                            fontSize: 14, fontWeight: FontWeight.w600)),
                    Text(item.createdAt,
                        style: TextStyle(
                            fontSize: 11, color: Theme.of(context).hintColor)),
                  ],
                ),
              ),
              Container(
                padding:
                    const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                decoration: BoxDecoration(
                  color: Theme.of(context).colorScheme.primary.withValues(alpha: 0.12),
                  borderRadius: BorderRadius.circular(12),
                ),
                child: Text(item.status, style: const TextStyle(fontSize: 11)),
              ),
            ],
          ),
          const Divider(height: 16),
          KeyValueRow('체인', '${item.chainId} (${item.chainType})'),
          KeyValueRow('보낸 주소', _shortHex(item.fromAddress)),
          if (item.toAddress != null) KeyValueRow('받는 주소', _shortHex(item.toAddress!)),
          KeyValueRow('txId', item.transactionId),
          if (item.txHash != null) ...[
            const SizedBox(height: 4),
            Text('txHash',
                style: TextStyle(fontSize: 11, color: Theme.of(context).hintColor)),
            CopyableText(item.txHash!),
          ],
        ],
      ),
    );
  }
}

String _shortHex(String s) =>
    s.length <= 16 ? s : '${s.substring(0, 10)}…${s.substring(s.length - 4)}';
